src/inc/til/bitmap.h

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#ifdef UNIT_TESTING
+class BitmapTests;
+#endif
+
+namespace til // Terminal Implementation Library. Also: "Today I Learned"
+{
+    class _bitmap_const_iterator
+    {
+    public:
+        _bitmap_const_iterator(const std::vector<bool>& values, til::rectangle rc, ptrdiff_t pos) :
+            _values(values),
+            _rc(rc),
+            _pos(pos),
+            _end(rc.size().area())
+        {
+            _calculateArea();
+        }
+
+        _bitmap_const_iterator& operator++()
+        {
+            _pos = _nextPos;
+            _calculateArea();
+            return (*this);
+        }
+
+        bool operator==(const _bitmap_const_iterator& other) const
+        {
+            return _pos == other._pos && _values == other._values;
+        }
+
+        bool operator!=(const _bitmap_const_iterator& other) const
+        {
+            return !(*this == other);
+        }
+
+        bool operator<(const _bitmap_const_iterator& other) const
+        {
+            return _pos < other._pos;
+        }
+
+        bool operator>(const _bitmap_const_iterator& other) const
+        {
+            return _pos > other._pos;
+        }
+
+        til::rectangle operator*() const
+        {
+            return _run;
+        }
+
+    private:
+        const std::vector<bool>& _values;
+        const til::rectangle _rc;
+        ptrdiff_t _pos;
+        ptrdiff_t _nextPos;
+        const ptrdiff_t _end;
+        til::rectangle _run;
+
+        void _calculateArea()
+        {
+            // Backup the position as the next one.
+            _nextPos = _pos;
+
+            // Seek forward until we find an on bit.
+            while (_nextPos < _end && !_values.at(_nextPos))
+            {
+                ++_nextPos;
+            }
+
+            // If we haven't reached the end yet...
+            if (_nextPos < _end)
+            {
+                // pos is now at the first on bit.
+                const auto runStart = _rc.point_at(_nextPos);
+
+                // We'll only count up until the end of this row.
+                // a run can be a max of one row tall.
+                const ptrdiff_t rowEndIndex = _rc.index_of(til::point(_rc.right() - 1, runStart.y())) + 1;
+
+                // Find the length for the rectangle.
+                ptrdiff_t runLength = 0;
+
+                // We have at least 1 so start with a do/while.
+                do
+                {
+                    ++_nextPos;
+                    ++runLength;
+                } while (_nextPos < _end && _nextPos < rowEndIndex && _values.at(_nextPos));

```

Comment:
logic could be simplified by `std::min`ing rowEndIndex with _end and saving that off as rowEndIndex.

Or, alternatively, the last row will only _ever_ end at the end of the buffer; is the _end check superfluous?

Reply:
Yeah it is. Removed superfluous check.